engine/classes/Elgg/EmailService.php

```diff
@@ -142,5 +150,185 @@ public function transport(Email $email) {
 
 		return true;
 	}
+	
+	/**
+	 * Build the body part of the e-mail message
+	 *
+	 * @param Email $email Email
+	 *
+	 * @return \Zend\Mime\Message
+	 */
+	protected function buildMessageBody(Email $email) {
+		// create body
+		$body = new MimeMessage();
+		
+		// add plain text part
+		$plain_text = elgg_strip_tags($email->getBody());
+		$plain_text = html_entity_decode($plain_text, ENT_QUOTES, 'UTF-8');
+		$plain_text = wordwrap($plain_text);
+		
+		$plain_text_part = new Part($plain_text);
+		$plain_text_part->setId('plaintext');
+		$plain_text_part->setType(Mime::TYPE_TEXT);
+		$plain_text_part->setCharset('UTF-8');
+		
+		$body->addPart($plain_text_part);
+		
+		// process attachements
+		$attachements = $this->processAttachements($email);
+		foreach ($attachements as $attachement) {
+			$body->addPart($attachement);
+		}
+		
+		return $body;
+	}
+	
+	/**
+	 * Extract the attachments from the e-mail and make them into parts for the
+	 * e-mail body
+	 *
+	 * @param Email $email Email
+	 *
+	 * @return \Zend\Mime\Part[]
+	 */
+	protected function processAttachements(Email $email) {
+		
+		$email_params = $email->getParams();
+		if (empty($email_params)) {
+			return [];
+		}
+		
+		// params added to the Email
+		// @see elgg_send_email()
+		$attachments = elgg_extract('attachments', $email_params, []);
+		
+		// check for params gotten through the notification system
+		// eg. notify_user or a notification event
+		$notification = elgg_extract('notification', $email_params);
+		if ($notification instanceof Notification) {
+			$notification_params = $notification->params;
+			if (!empty($notification_params)) {
+				$notification_attachments = elgg_extract('attachments', $notification_params, []);
+				
+				foreach ($notification_attachments as $attachment) {
+					$attachments[] = $attachment;
+				}
+			}
+		}
+		
+		if (empty($attachments)) {
+			return [];
+		}
+		
+		$parts = [];
+		foreach ($attachments as $index => $attachment) {
+			$part = false;
+			
+			if ($attachment instanceof Part) {
+				$part = $attachment;
+			}
+			
+			if ($attachment instanceof \ElggFile) {
+				$part = $this->elggFileToAttachement($attachment);
+			}
+			
+			if (is_array($attachment)) {
+				$part = $this->arrayToAttachement($attachment);
+			}
+			
+			if (empty($part)) {
+				continue;
+			}
+			
+			if (!isset($part->id)) {
+				$part->setId("attachment_{$index}");
+			}
+			
+			if (!isset($part->filename)) {
+				$part->setFileName("file_{$index}");
+			}
+			
+			if (!isset($part->disposition)) {
+				$part->setDisposition('attachment');
+			}
+			
+			$parts[] = $part;
+		}
+		
+		return $parts;
+	}
+	
+	/**
+	 * Make an ElggFile into a correct formatted attachment
+	 *
+	 * @param \ElggFile $file the ElggFile to process
+	 *
+	 * @return false|\Zend\Mime\Part
+	 */
+	protected function elggFileToAttachement(\ElggFile $file) {
```

Comment:
I don't think this belongs here, make some utility/adapter class. Better yet add ElggEmailAttachment interface that extends Part and require that attachments comply with the instance. I don't like all this boilerplate here

Reply:
Maybe in the Email class?

Reply:
I'd say add `EmailAttachment` class with static `factory` methods. Then you could just use `Email::attach(EmailAttachment)`. Somewhere along the lines, add a hook to filter attachments.